src/cpp/session/modules/SessionRmdNotebook.R

```diff
@@ -614,142 +615,197 @@ assign(".rs.notebookVersion", envir = .rs.toolsEnv(), "1.0")
-  matches <- unlist(regmatches(rmdChunkOpts, regexec(.rs.reRmdChunkBegin(), rmdChunkOpts)))
-  if (length(matches) > 1)
-    rmdChunkOpts <- matches[[2]]
-
-  tryCatch({
-    # if this is the setup chunk, it's not included by default
-    setupIndicator <- "r setup"
-    if (identical(substring(rmdChunkOpts, 1, nchar(setupIndicator)),
-                  setupIndicator)) {
-      opts$include <- FALSE
-    }
-
-    # extract and remove engine name (can be overridden below with engine=)
-    opts$engine <- unlist(strsplit(rmdChunkOpts, split = "(\\s|,)+"))[[1]]
-    rmdChunkOpts <- substring(rmdChunkOpts, nchar(opts$engine) + 1)
-
-    # parse them, then merge with the defaults (evaluate in global environment)
-    opts <- .rs.mergeLists(opts,
-                           eval(substitute(knitr:::parse_params(rmdChunkOpts)),
-                                envir = .GlobalEnv))
-
-    # if a chunk option is provided in both body (yaml chunk options) and header (rmdChunkOpts)
-    # the body should override the header option
-    opts <- opts[unique(names(opts))]
-    
-    # convert T, F for R code chunk options to TRUE, FALSE as appropriate
-    opts <- lapply(opts, function(opt) {
-      if (identical(opt, as.name("T")))
-        TRUE
-      else if (identical(opt, as.name("F")))
-        FALSE
-      else
-        opt
-    })
+   # remove comment prefix
+   commentPrefixPattern <- sprintf("^%s\\|\\s*", commentPrefix)
+   opt <- sub(commentPrefixPattern, "", opt)
+   
+   # check for YAML entry, mimicing code in knitr
+   if (grepl("^[^ :]+:($|\\s)", opt[[1]])) {
+      meta <- .rs.tryCatch(.rs.parseYamlOptImpl(opt))
+      if (!inherits(meta, "error"))
+         return(meta)
+   }
+   
+   # fallback to regular params parser
+   opts <- paste(opt, collapse = " ")
+   knitr:::parse_params(opts, label = FALSE)
+})
 
-    # convert language name objects to plain characters (these can occur in
-    # malformed expressions, and cause scalar conversion below to fail)
-    names <- vapply(opts, is.name, TRUE)
-    opts[names] <- as.character(opts[names])
-    
-    # ensure that fields we expect to be logical are actually logical
-    # (sanitize invalid inputs here since they can break notebook execution)
-    fields <- list(
+.rs.addFunction("evaluateChunkOptions", function(id, type, code)
+{
+   Encoding(code) <- "UTF-8"
+   code <- unlist(strsplit(code, "\n", fixed = TRUE))
+   
+   # parsed chunk options; start with default chunk options
+   opts <- list()
+   
+   # first, parse chunk options embedded in the chunk header
+   rmdChunkOpts <- code[[1]]
+   
+   # strip chunk indicators if present
+   matches <- unlist(regmatches(rmdChunkOpts, regexec(.rs.reRmdChunkBegin(), rmdChunkOpts)))
+   if (length(matches) > 1)
+      rmdChunkOpts <- matches[[2]]
+   
+   # attempt to parse parameters (swallow errors)
+   .rs.tryCatch({
+      
+      # if this is the setup chunk, it's not included by default
+      if (.rs.startsWith(rmdChunkOpts, "r setup"))
```

Comment:
just realizing, couldn't the setup chunk also be `r, setup`? it seems like we don't check for the comma?

Reply:
Good idea -- instead of checking the chunk text itself, we should just check if it has a `setup` label.